pulp_ostree/app/viewsets.py

```diff
@@ -196,14 +454,72 @@ class Meta:
         fields = {"name": NAME_FILTER_OPTIONS}
 
 
-class OstreeRefViewSet(ReadOnlyContentViewSet):
+class OstreeContentQuerySetMixin:
+    """
+    A mixin that filters content units based on their object-level permissions.
+    """
+
+    def _scope_repos_by_repo_version(self, repo_version_href):
+        repo_version = core.NamedModelViewSet.get_resource(repo_version_href, RepositoryVersion)
+        repo = repo_version.repository.cast()
+
+        has_model_perm = self.request.user.has_perm(REPO_VIEW_PERM)
+        has_object_perm = self.request.user.has_perm(REPO_VIEW_PERM, repo)
+
+        if has_model_perm or has_object_perm:
+            return [repo]
+        else:
+            return []
+
+    def get_content_qs(self, qs):
+        """
+        Get a filtered QuerySet based on the current request's scope.
+
+        This method returns only content units a user is allowed to preview. The user with the
+        global import and mirror permissions (i.e., having the "ostree.view_ostreerepository")
+        can see orphaned content too.
+        """
+        if self.request.user.has_perm(REPO_VIEW_PERM):
+            return qs
+
+        if repo_version_href := self.request.query_params.get("repository_version"):
+            allowed_repos = self._scope_repos_by_repo_version(repo_version_href)
+        else:
+            allowed_repos = get_objects_for_user(
+                self.request.user, REPO_VIEW_PERM, models.OstreeRepository.objects.all()
+            ).only("pk")
+
+        return qs.model.objects.filter(repositories__in=allowed_repos)
+
+
+class OstreeRefViewSet(OstreeContentQuerySetMixin, ReadOnlyContentViewSet):
     """A ViewSet class for OSTree head commits."""
 
     endpoint_name = "refs"
     queryset = models.OstreeRef.objects.all()
     serializer_class = serializers.OstreeRefSerializer
     filterset_class = OstreeRefFilter
 
+    DEFAULT_ACCESS_POLICY = {
+        "statements": [
+            {
+                "action": ["list", "retrieve"],
+                "principal": "authenticated",
+                "effect": "allow",
+            },
+            {
+                "action": ["create"],
+                "principal": "authenticated",
+                "effect": "allow",
+                "condition": [
+                    "has_required_repo_perms_on_upload:ostree.modify_ostreerepository",
+                    "has_upload_param_model_or_obj_perms:core.change_upload",
```

Comment:
We cannot upload refs to Pulp, can we?

Other plugins have content viewsets which inherit from, e.g., "SingleArtifactContentUploadViewSet". I am not sure if we do the same here. Can you verify that, please?